plugins/libplugin-pay.c

```diff
@@ -1606,3 +1607,170 @@ static void exemptfee_cb(struct exemptfee_data *d, struct payment *p)
+					       const char *buf,
+					       const jsmntok_t *result,
+					       struct payment *p)
+{
+	/* Use reservoir sampling across the capable channels. */
+	struct shadow_route_data *d = payment_mod_shadowroute_get_data(p);
+	struct payment_constraints *cons = &d->constraints;
+	struct route_info *best = NULL;
+	size_t i;
+	u64 sample = 0;
+	struct amount_msat best_fee;
+	const jsmntok_t *sattok, *delaytok, *basefeetok, *propfeetok, *desttok,
+	    *channelstok, *chan;
+
+	channelstok = json_get_member(buf, result, "channels");
+	json_for_each_arr(i, chan, channelstok) {
+		u64 v = pseudorand(UINT64_MAX);
+		struct route_info curr;
+		struct amount_sat capacity;
+		struct amount_msat fee;
+
+		sattok = json_get_member(buf, chan, "satoshis");
+		delaytok = json_get_member(buf, chan, "delay");
+		basefeetok = json_get_member(buf, chan, "base_fee_millisatoshi");
+		propfeetok = json_get_member(buf, chan, "fee_per_millionth");
+		desttok =  json_get_member(buf, chan, "destination");
+
+		if (sattok == NULL || delaytok == NULL ||
+		    delaytok->type != JSMN_PRIMITIVE || basefeetok == NULL ||
+		    basefeetok->type != JSMN_PRIMITIVE || propfeetok == NULL ||
+		    propfeetok->type != JSMN_PRIMITIVE || desttok == NULL)
+			continue;
+
+		json_to_u16(buf, delaytok, &curr.cltv_expiry_delta);
+		json_to_number(buf, basefeetok, &curr.fee_base_msat);
+		json_to_number(buf, propfeetok,
+			       &curr.fee_proportional_millionths);
+		json_to_sat(buf, sattok, &capacity);
+		json_to_node_id(buf, desttok, &curr.pubkey);
+
+		if (!best || v > sample) {
+			/* If the capacity is insufficient to pass the amount
+			 * it's not a plausible extension. */
+			if (amount_msat_greater_sat(p->amount, capacity))
+				continue;
+
+			if (curr.cltv_expiry_delta > cons->cltv_budget)
+				continue;
+
+			if (!amount_msat_fee(
+				&fee, p->amount, curr.fee_base_msat,
+				curr.fee_proportional_millionths)) {
+				/* Fee computation failed... */
+				continue;
+			}
+
+			if (amount_msat_greater_eq(fee, cons->fee_budget))
+				continue;
+
+			best = tal_dup(tmpctx, struct route_info, &curr);
+			best_fee = fee;
+			sample = v;
+		}
+	}
+
+	if (best != NULL) {
+		bool ok;
+		/* Ok, we found an extension, let's add it and roll the dice again. */
+		d->destination = best->pubkey;
+		/* Apply deltas to the constraints in the shadow route so we
+		 * don't overshoot our 1/4th target. */
+		ok = payment_constraints_update(&d->constraints, best_fee,
+						best->cltv_expiry_delta);
+		/* Now do the same to the payment constraints so other
+		 * modifiers don't do it either. */
+		ok &= payment_constraints_update(&p->constraints, best_fee,
+						 best->cltv_expiry_delta);
+
+		/* And now the thing that caused all of this: adjust the call
+		 * to getroute. */
+		ok &= amount_msat_add(&p->getroute->amount, p->getroute->amount,
+				      best_fee);
+		p->getroute->cltv += best->cltv_expiry_delta;
+		assert(ok);
```

Comment:
Hmm, this seems like if we the shadow route gets too long, we assert()?  Shouldn't we stop (set best = NULL and fall through?).

Reply:
Also, what happens if our shadow route makes our onion too large?

Reply:
I added an early escape to continue with the next iteration if `payment_update_constraints` on the modifier (which is guaranteed to have 1/4th of the overall constraints). If that update fails we've exhausted our shadow budget and need to continue right away. Since the first constraint update succeeded (which has <= 1/4th of the overall constraints) we're guarded here, and the assert is just here to express that logic.

The onion is not touched at all by the shadow route (they are just amount and cltv offsets to the `getroute` calls, which then is what gets turned into the onion). So we could technically have an arbitrarily long shadow route, and the onion would not explode :-)

Reply:
Good point!